core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisOutputFormat.java

```diff
@@ -0,0 +1,166 @@
+    Configuration conf = context.getConfiguration();
+    BasicAWSCredentials awsCred = new BasicAWSCredentials(conf.get(Properties.KinesisRealtimeSink.ACCESS_ID),
+                                                          conf.get(Properties.KinesisRealtimeSink.ACCESS_KEY));
+
+    if (kinesisClient == null) {
+      LOG.info("Creating amazon kinesis client for stream {}", Properties.KinesisRealtimeSink.NAME);
+      kinesisClient = new AmazonKinesisClient(awsCred);
+      try {
+        KinesisUtil.createAndWaitForStream(kinesisClient, conf.get(Properties.KinesisRealtimeSink.NAME),
+                                           Integer.valueOf(conf.get(Properties.KinesisRealtimeSink.SHARD_COUNT)));
+      } catch (AmazonServiceException e) {
+        if (e.getStatusCode() == 400 && e.getErrorCode().equalsIgnoreCase("UnrecognizedClientException")) {
+          throw new IllegalArgumentException("AWS credentials are not correct", e);
+        } else {
+          throw e;
+        }
+      }
+    }
+  }
+
+  /**
+   * Get the output committer for this output format. This is responsible
+   * for ensuring the output is committed correctly.
+   *
+   * @param context the task context
+   * @return an output committer
+   * @throws IOException
+   * @throws InterruptedException
+   */
+  @Override
+  public OutputCommitter getOutputCommitter(TaskAttemptContext context) throws IOException, InterruptedException {
+    return new NoOpOutputCommitter();
+  }
+
+  private class KinesisRecordWriter extends RecordWriter<NullWritable, Text> {
+    private final AmazonKinesisClient kinesisClient;
+    private final TaskAttemptContext context;
+    private final boolean distribute;
+
+    KinesisRecordWriter(AmazonKinesisClient kinesisClient, TaskAttemptContext context) {
+      this.kinesisClient = kinesisClient;
+      this.context = context;
+      distribute = "true".equals(context.getConfiguration().get(Properties.KinesisRealtimeSink.DISTRIBUTE));
+    }
+
+    /**
+     * Writes a key/value pair.
+     *
+     * @param key   the key to write.
+     * @param value the value to write.
+     * @throws IOException
+     */
+    @Override
+    public void write(NullWritable key, Text value) throws IOException, InterruptedException {
+      PutRecordRequest putRecordRequest = new PutRecordRequest();
+      putRecordRequest.setStreamName(context.getConfiguration().get(Properties.KinesisRealtimeSink.NAME));
+      // uniformly distribute data between shards if distribute is set to true, Otherwise write everything to a
+      // single shard
+      String dist = distribute ? String.valueOf(Math.random()) : String.valueOf(distribute);
+      putRecordRequest.setPartitionKey(dist);
+      ByteBuffer buffer = ByteBuffer.wrap(value.toString().getBytes());
+      sendKinesisRecord(buffer, putRecordRequest);
+    }
+
+    private void sendKinesisRecord(ByteBuffer data, PutRecordRequest putRecordRequest) {
+      int retry = 0;
+      while (true) {
```

Comment:
I have a feeling that a `for (int i = 0; i < MAX_RETRIES; i++) {` loop will look safer here.
I know this while loop will only iterate up to MAX_RETRIES times, but upon first glance, it can look like there's a risk of infinite-while loop here.